File: plugins/php/sample_plugin.php
#!/usr/bin/env php
<?php
/**
 * sample_plugin.php
 *
 * Production-ready PHP plugin for OmniFlow (TheSkiF4er/OmniFlow)
 * License: Apache-2.0
 *
 * Overview:
 *  - CLI script that communicates with the host via newline-delimited JSON over stdin/stdout.
 *  - Robust input size limiting, structured logging to stderr (optional JSON), graceful shutdown,
 *    heartbeat background behavior, configurable env vars, and simple exec-timeout support where available.
 *  - Designed to be portable across typical PHP CLI environments. Uses only core PHP functions where possible.
 *
 * Runtime contract (newline-delimited JSON):
 *  Host -> Plugin:
 *    { "id": "<uuid>", "type": "exec|health|shutdown", "payload": {...} }
 *
 *  Plugin -> Host responses:
 *    { "id":"<uuid>", "status":"ok|error", "code"?:int, "message"?:string, "body"?:object }
 *
 * Environment variables (recommended):
 *  - OMNIFLOW_PLUGIN_MAX_LINE (default 131072)   # max bytes per incoming message
 *  - OMNIFLOW_PLUGIN_HEARTBEAT (default 5)       # heartbeat seconds
 *  - OMNIFLOW_LOG_JSON (default empty)           # if set, emit JSON logs to stderr
 *  - OMNIFLOW_EXEC_TIMEOUT (default 10)          # seconds timeout for exec handlers (best-effort)
 *
 * Build/Run:
 *  - Requires PHP 7.4+ (recommended 8.x) with CLI SAPI.
 *  - Run: echo '{"id":"1","type":"health"}' | ./sample_plugin.php
 *
 * Notes on timeouts:
 *  - If the pcntl extension is available, this script will attempt to use pcntl_alarm for exec timeouts
 *    and pcntl_async_signals for responsive shutdown. If pcntl is not available, exec timeout is best-effort
 *    and relies on the handler doing non-blocking work.
 */

// ---------------- Configuration ----------------
$MAX_LINE = getenv('OMNIFLOW_PLUGIN_MAX_LINE') !== false ? (int)getenv('OMNIFLOW_PLUGIN_MAX_LINE') : 131072;
$HEARTBEAT = getenv('OMNIFLOW_PLUGIN_HEARTBEAT') !== false ? (int)getenv('OMNIFLOW_PLUGIN_HEARTBEAT') : 5;
$LOG_JSON = getenv('OMNIFLOW_LOG_JSON') !== false;
$EXEC_TIMEOUT = getenv('OMNIFLOW_EXEC_TIMEOUT') !== false ? (int)getenv('OMNIFLOW_EXEC_TIMEOUT') : 10;

define('PLUGIN_NAME', 'OmniFlowPHPRelease');
define('PLUGIN_VERSION', '1.0.0');

// runtime control
$running = true; // when false, main loop exits
$shutdown_requested = false;

// track last heartbeat time (monotonic)
$lastHeartbeat = hrtime(true);

// Try to enable async signals (PHP 7.1+ with pcntl)
$pcntlAvailable = function_exists('pcntl_async_signals') && function_exists('pcntl_signal');
if ($pcntlAvailable) {
    pcntl_async_signals(true);
    pcntl_signal(SIGINT, function($sig) use (&$running, &$shutdown_requested) {
        stderr_log('WARN', "received SIGINT, initiating shutdown");
        $shutdown_requested = true;
        $running = false;
    });
    pcntl_signal(SIGTERM, function($sig) use (&$running, &$shutdown_requested) {
        stderr_log('WARN', "received SIGTERM, initiating shutdown");
        $shutdown_requested = true;
        $running = false;
    });
}

// If pcntl available, also use alarm for timeouts (best-effort)
$pcntlAlarmAvailable = function_exists('pcntl_alarm') && $pcntlAvailable;

// ---------------- Logging ----------------
function current_time_iso8601() {
    return gmdate('Y-m-d\TH:i:s\Z');
}

function stderr_log($level, $message, $extra = null) {
    global $LOG_JSON;
    $time = current_time_iso8601();
    if ($LOG_JSON) {
        $obj = ['time' => $time, 'level' => $level, 'plugin' => PLUGIN_NAME, 'message' => $message];
        if ($extra !== null) $obj['extra'] = $extra;
        fwrite(STDERR, json_encode($obj, JSON_UNESCAPED_SLASHES | JSON_UNESCAPED_UNICODE) . "\n");
    } else {
        fwrite(STDERR, "$time [$level] " . PLUGIN_NAME . ": $message\n");
    }
}

function info_log($msg, $extra = null) { stderr_log('INFO', $msg, $extra); }
function warn_log($msg, $extra = null) { stderr_log('WARN', $msg, $extra); }
function error_log_msg($msg, $extra = null) { stderr_log('ERROR', $msg, $extra); }

// ---------------- Respond helpers ----------------
function respond($data) {
    // Ensure newline terminated, single-line JSON
    echo json_encode($data, JSON_UNESCAPED_SLASHES | JSON_UNESCAPED_UNICODE) . "\n";
    flush();
}

function respond_ok($id, $body = null) {
    $r = ['status' => 'ok'];
    if ($id !== null) $r['id'] = $id;
    if ($body !== null) $r['body'] = $body;
    respond($r);
}

function respond_error($id, $code, $message) {
    $r = ['status' => 'error', 'code' => $code, 'message' => $message];
    if ($id !== null) $r['id'] = $id;
    respond($r);
}

// ---------------- Utility: safe read line with limit ----------------
function safe_read_line($maxLine) {
    // Use stream_get_line to limit bytes read (includes delimiter)
    $line = @stream_get_line(STDIN, $maxLine + 1, "\n");
    if ($line === false) {
        // EOF or error
        if (feof(STDIN)) return null;
        return false;
    }
    // stream_get_line removes the delimiter; we have the line content
    $bytes = strlen($line);
    if ($bytes > $maxLine) {
        return ['error' => 'too_large'];
    }
    return $line;
}

// ---------------- Exec timeout helpers (best-effort) ----------------
$alarmTriggered = false;
function alarm_handler($sig) {
    global $alarmTriggered;
    $alarmTriggered = true;
}
if ($pcntlAlarmAvailable) {
    pcntl_signal(SIGALRM, 'alarm_handler');
}

// ---------------- Action handlers ----------------
function handle_health($id) {
    $body = ['status' => 'healthy', 'version' => PLUGIN_VERSION];
    respond_ok($id, $body);
}

function handle_exec($id, $payload) {
    global $EXEC_TIMEOUT, $pcntlAlarmAvailable, $alarmTriggered;
    // payload expected as array/object
    if (!is_array($payload) && !is_object($payload)) {
        respond_error($id, 400, "missing or invalid payload");
        return;
    }
    // ensure it's associative
    $action = null;
    if (is_array($payload) && array_key_exists('action', $payload)) $action = $payload['action'];
    elseif (is_object($payload) && property_exists($payload, 'action')) $action = $payload->action;

    if (!is_string($action)) { respond_error($id, 400, "missing or invalid 'action'"); return; }

    // Best-effort timeout using pcntl_alarm
    $alarmTriggered = false;
    if ($pcntlAlarmAvailable) {
        pcntl_alarm($EXEC_TIMEOUT);
    }

    // Implement safe, quick built-in actions: echo, reverse, compute
    try {
        if ($action === 'echo') {
            $message = '';
            if (is_array($payload) && isset($payload['message'])) $message = (string)$payload['message'];
            if (is_object($payload) && property_exists($payload, 'message')) $message = (string)$payload->message;
            $body = ['action' => 'echo', 'message' => $message];
            respond_ok($id, $body);
        } elseif ($action === 'reverse') {
            $message = '';
            if (is_array($payload) && isset($payload['message'])) $message = (string)$payload['message'];
            if (is_object($payload) && property_exists($payload, 'message')) $message = (string)$payload->message;
            // reverse multibyte-safe
            $rev = implode('', array_reverse(preg_split('//u', $message, -1, PREG_SPLIT_NO_EMPTY)));
            $body = ['action' => 'reverse', 'message' => $rev];
            respond_ok($id, $body);
        } elseif ($action === 'compute') {
            // sum numbers
            $numbers = null;
            if (is_array($payload) && isset($payload['numbers'])) $numbers = $payload['numbers'];
            if (is_object($payload) && property_exists($payload, 'numbers')) $numbers = $payload->numbers;
            if (!is_array($numbers)) { respond_error($id, 400, "missing or invalid 'numbers' array"); return; }
            $sum = 0.0;
            foreach ($numbers as $n) {
                if (!is_numeric($n)) { respond_error($id, 400, "numbers must be numeric"); return; }
                $sum += (float)$n;
            }
            $body = ['action' => 'compute', 'sum' => $sum];
            respond_ok($id, $body);
        } else {
            respond_error($id, 422, 'unsupported action');
        }
    } catch (Throwable $e) {
        // If alarm triggered, send timeout
        if ($alarmTriggered) {
            respond_error($id, 408, 'exec timeout');
        } else {
            respond_error($id, 500, 'internal error');
            error_log_msg('exec handler exception: ' . $e->getMessage());
        }
    } finally {
        if ($pcntlAlarmAvailable) pcntl_alarm(0); // cancel alarm
    }
}

// ---------------- Main loop ----------------
info_log("starting plugin version=" . PLUGIN_VERSION . " max_line={$MAX_LINE} heartbeat={$HEARTBEAT}");

while ($running) {
    // heartbeat management (non-blocking): log heartbeat every HEARTBEAT seconds
    $now = hrtime(true);
    $elapsedSec = ($now - $lastHeartbeat) / 1e9;
    if ($elapsedSec >= $HEARTBEAT) {
        $lastHeartbeat = $now;
        info_log("heartbeat");
    }

    // Read line with limit; stream_get_line will block until delimiter or EOF
    $res = safe_read_line($MAX_LINE);
    if ($res === null) {
        // EOF
        info_log("stdin closed (EOF), exiting");
        break;
    }
    if ($res === false) {
        warn_log("error reading stdin");
        // small sleep to avoid tight loop on I/O error
        usleep(100000);
        continue;
    }
    if (is_array($res) && isset($res['error']) && $res['error'] === 'too_large') {
        warn_log("incoming message too large, rejecting");
        respond_error(null, 413, 'payload too large');
        // drain remainder of line (already handled by stream_get_line limit)
        continue;
    }

    $line = $res;
    $trim = trim($line);
    if ($trim === '') continue;

    // Decode JSON
    $msg = json_decode($line, true);
    if ($msg === null || !is_array($msg)) {
        warn_log('invalid JSON message');
        respond_error(null, 400, 'invalid JSON');
        continue;
    }

    $id = array_key_exists('id', $msg) ? $msg['id'] : null;
    $type = array_key_exists('type', $msg) ? strtolower((string)$msg['type']) : null;
    $payload = array_key_exists('payload', $msg) ? $msg['payload'] : null;

    if ($type === null) { respond_error($id, 400, "missing 'type'"); continue; }

    if ($type === 'health') {
        handle_health($id);
        continue;
    }
    if ($type === 'exec') {
        handle_exec($id, $payload);
        continue;
    }
    if ($type === 'shutdown' || $type === 'quit') {
        respond_ok($id, ['result' => 'shutting_down']);
        $shutdown_requested = true;
        $running = false;
        break;
    }

    respond_error($id, 400, 'unknown type');
}

// Graceful shutdown cleanup
info_log('plugin shutdown complete');
exit(0);

?>

File: plugins/c/sample_plugin.c
/*
 * sample_plugin.c
 *
 * Sample C plugin for OmniFlow (TheSkiF4er/OmniFlow)
 * License: Apache-2.0
 *
 * Purpose:
 *   - Demonstrates a secure, production-minded template for writing a plugin
 *     in C that integrates with the OmniFlow plugin host via a simple
 *     JSON-over-stdin/stdout RPC-style protocol.
 *   - Shows secure initialization, configuration via environment variables,
 *     structured logging, graceful shutdown, health checks, background tasks,
 *     and careful resource management.
 *
 * Build (recommended):
 *   gcc -std=c11 -O2 -Wall -Wextra -pthread -o sample_plugin sample_plugin.c
 *
 * Security notes:
 *   - Avoids dynamic code evaluation and uses limited parsing to avoid
 *     large dependency surface.
 *   - Uses timeouts and resource limits where appropriate.
 *   - Logs to stderr; protocol responses go to stdout.
 *
 * Integration contract (example):
 *   - Host sends JSON messages to plugin's stdin, terminated by newline.
 *   - Plugin writes JSON responses to stdout, terminated by newline.
 *   - Messages have simple shape: {"id": "<id>", "type":"exec|health|quit", "payload": {...}}
 *
 * Replace/extend parsing with a robust JSON library in production (cJSON, jsmn, sajson).
 */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <stdbool.h>
#include <pthread.h>
#include <signal.h>
#include <stdatomic.h>
#include <time.h>
#include <errno.h>
#include <unistd.h>

/* ---------------- Configuration & Constants ---------------- */

#define PLUGIN_NAME "OmniFlowSampleC"
#define PLUGIN_VERSION "0.1.0"
#define MAX_LINE 65536
#define LOG_PREFIX "[omni-plugin:c]"

/* Read env var: OMNIFLOW_PLUGIN_TIMEOUT (seconds) */
static int plugin_timeout_seconds = 10;

/* Background worker control */
static atomic_bool running = true;
static pthread_t bg_thread;

/* ---------------- Utilities: Logging, time ---------------- */

static void safe_log(const char *fmt, ...)
    __attribute__((format(printf, 1, 2)));

static void safe_log(const char *fmt, ...)
{
    va_list ap;
    va_start(ap, fmt);
    fprintf(stderr, "%s %s: ", LOG_PREFIX, PLUGIN_NAME);
    vfprintf(stderr, fmt, ap);
    fprintf(stderr, "\n");
    fflush(stderr);
    va_end(ap);
}

static void current_time_iso8601(char *buf, size_t len)
{
    time_t t = time(NULL);
    struct tm tm;
    gmtime_r(&t, &tm);
    strftime(buf, len, "%Y-%m-%dT%H:%M:%SZ", &tm);
}

/* ---------------- Tiny JSON helpers (safe, limited) ---------------- */
/*
 * These helpers are intentionally minimal: they search for string keys and
 * extract simple string or integer values. For production use, link a proper
 * JSON parser (cJSON, jsmn, yajl). The limited parser reduces dependency
 * surface for an example plugin.
 */

/* Find a string value for key in JSON (very limited):
 *   json: full JSON text (null-terminated)
 *   key: plain key name (without quotes)
 *   out: buffer to fill
 *   out_len: capacity
 * Returns true on success. */
static bool json_get_string(const char *json, const char *key, char *out, size_t out_len)
{
    if (!json || !key || !out) return false;
    size_t keylen = strlen(key);
    /* pattern "key"\s*:\s*"value" */
    const char *p = json;
    while ((p = strstr(p, "\"") ) != NULL) {
        /* p points to a quote, check if what follows is key" */
        if (strncmp(p+1, key, keylen) == 0 && p[1+keylen] == '"') {
            const char *colon = strchr(p+1+keylen+1, ':');
            if (!colon) { p += 1; continue; }
            /* find opening quote for value */
            const char *valq = strchr(colon, '"');
            if (!valq) { p += 1; continue; }
            valq++; /* move after opening quote */
            const char *valq_end = valq;
            while (*valq_end && (*valq_end != '"' || *(valq_end-1) == '\\')) valq_end++;
            size_t vlen = valq_end - valq;
            if (vlen >= out_len) vlen = out_len - 1;
            memcpy(out, valq, vlen);
            out[vlen] = '\0';
            return true;
        }
        p += 1;
    }
    return false;
}

/* Find integer value for key. Very limited: handles optional spaces and digits */
static bool json_get_int(const char *json, const char *key, long *out)
{
    if (!json || !key || !out) return false;
    size_t keylen = strlen(key);
    const char *p = json;
    while ((p = strstr(p, "\"") ) != NULL) {
        if (strncmp(p+1, key, keylen) == 0 && p[1+keylen] == '"') {
            const char *colon = strchr(p+1+keylen+1, ':');
            if (!colon) { p += 1; continue; }
            const char *q = colon + 1;
            while (*q && (*q == ' ' || *q == '\t')) q++;
            char *endptr = NULL;
            long v = strtol(q, &endptr, 10);
            if (q == endptr) return false;
            *out = v;
            return true;
        }
        p += 1;
    }
    return false;
}

/* ---------------- Protocol helpers ---------------- */

/* Write JSON response to stdout and flush. Uses a simple protocol:
 * {"id":"...","response":{...}}
 */
static void respond_ok(const char *id, const char *body_json)
{
    if (!id) id = "";
    if (!body_json) body_json = "{}";
    /* Ensure single-line safe output */
    printf("{\"id\":\"%s\",\"status\":\"ok\",\"time\":\"%s\",\"body\":%s}\n",
           id, ({ char _t[32]; current_time_iso8601(_t, sizeof(_t)); strdup(_t); }),
           body_json);
    fflush(stdout);
}

static void respond_error(const char *id, int code, const char *message)
{
    if (!id) id = "";
    if (!message) message = "unknown error";
    printf("{\"id\":\"%s\",\"status\":\"error\",\"code\":%d,\"message\":\"%s\",\"time\":\"%s\"}\n",
           id, code, message, ({ char _t[32]; current_time_iso8601(_t, sizeof(_t)); strdup(_t); }));
    fflush(stdout);
}

/* Note: we used statement expressions above for brevity. If your compiler doesn't support
 * GNU extensions, replace with a small helper to call current_time_iso8601(). */

/* ---------------- Background worker (example) ---------------- */
static void *background_worker(void *arg)
{
    safe_log("background worker started");
    int counter = 0;
    while (atomic_load(&running)) {
        /* Example periodic work: emit heartbeat to stderr once per interval */
        sleep(5);
        counter++;
        safe_log("heartbeat %d", counter);
    }
    safe_log("background worker stopping");
    return NULL;
}

/* ---------------- Signal handling ---------------- */
static void handle_signal(int sig)
{
    safe_log("received signal %d, initiating shutdown", sig);
    atomic_store(&running, false);
}

/* ---------------- Plugin lifecycle ---------------- */

static void plugin_init(void)
{
    const char *t = getenv("OMNIFLOW_PLUGIN_TIMEOUT");
    if (t) {
        char *end;
        long v = strtol(t, &end, 10);
        if (end != t && v > 0 && v <= 3600) plugin_timeout_seconds = (int)v;
    }
    safe_log("init: version=%s timeout=%d", PLUGIN_VERSION, plugin_timeout_seconds);

    /* Setup signal handlers for graceful shutdown */
    struct sigaction sa;
    memset(&sa, 0, sizeof(sa));
    sa.sa_handler = handle_signal;
    sigaction(SIGINT, &sa, NULL);
    sigaction(SIGTERM, &sa, NULL);

    /* Start background worker */
    atomic_store(&running, true);
    if (pthread_create(&bg_thread, NULL, background_worker, NULL) != 0) {
        safe_log("failed to create background thread: %s", strerror(errno));
        atomic_store(&running, false);
    }
}

static void plugin_shutdown(void)
{
    safe_log("shutdown requested");
    atomic_store(&running, false);
    /* Join background thread with timeout */
    void *res;
    if (pthread_join(bg_thread, &res) != 0) {
        safe_log("failed to join background thread: %s", strerror(errno));
    }
    safe_log("shutdown complete");
}

/* ---------------- Command handlers ---------------- */

static void handle_health(const char *id)
{
    /* Return basic health payload */
    respond_ok(id, "{\"status\":\"healthy\",\"uptime_seconds\":0}");
}

static void handle_exec(const char *id, const char *payload)
{
    /* Example: payload may contain {"action":"echo","message":"..."}
     * We'll support a safe "echo" action and a "reverse" action for demo.
     */
    char action[128] = {0};
    char message[4096] = {0};
    if (!json_get_string(payload ? payload : "", "action", action, sizeof(action))) {
        respond_error(id, 400, "missing or invalid action");
        return;
    }
    if (!json_get_string(payload ? payload : "", "message", message, sizeof(message))) {
        /* message optional; default empty */
        message[0] = '\0';
    }

    if (strcmp(action, "echo") == 0) {
        /* Construct safe JSON body: we must escape quotes/backslashes in message; but for brevity
         * we'll limit message length and replace quotes. */
        for (size_t i = 0; i < strlen(message); ++i) {
            if (message[i] == '"' || message[i] == '\\') message[i] = '?';
        }
        char body[8192];
        snprintf(body, sizeof(body), "{\"action\":\"echo\",\"message\":\"%s\"}", message);
        respond_ok(id, body);
        return;
    }
    else if (strcmp(action, "reverse") == 0) {
        size_t n = strlen(message);
        char rev[4096];
        for (size_t i = 0; i < n && i < sizeof(rev)-1; ++i) rev[i] = message[n-1-i];
        rev[n < sizeof(rev)-1 ? n : (sizeof(rev)-1)] = '\0';
        for (size_t i = 0; i < strlen(rev); ++i) if (rev[i] == '"' || rev[i] == '\\') rev[i] = '?';
        char body[8192];
        snprintf(body, sizeof(body), "{\"action\":\"reverse\",\"message\":\"%s\"}", rev);
        respond_ok(id, body);
        return;
    }
    else {
        respond_error(id, 422, "unsupported action");
        return;
    }
}

/* ---------------- Main loop: read lines from stdin ---------------- */
int main(int argc, char **argv)
{
    (void)argc; (void)argv;

    /* Initialize plugin */
    plugin_init();

    char *line = NULL;
    size_t linecap = 0;
    ssize_t linelen;

    /* Set stdin to line-buffered and non-blocking timeout can be applied by select if needed */
    while (atomic_load(&running) && (linelen = getline(&line, &linecap, stdin)) != -1) {
        if (linelen == 0) continue;
        /* Trim newline */
        while (linelen > 0 && (line[linelen-1] == '\n' || line[linelen-1] == '\r')) { line[--linelen] = '\0'; }
        if (linelen == 0) continue;

        /* For safety, cap line length */
        if ((size_t)linelen > MAX_LINE) {
            safe_log("incoming line too long (%zd), rejecting", linelen);
            respond_error("", 413, "payload too large");
            continue;
        }

        /* Parse minimal fields: id and type */
        char id[128] = {0};
        char type[64] = {0};
        if (!json_get_string(line, "id", id, sizeof(id))) {
            /* not required: assign empty id */
            id[0] = '\0';
        }
        if (!json_get_string(line, "type", type, sizeof(type))) {
            respond_error(id, 400, "missing type");
            continue;
        }

        /* Extract payload substring (very small heuristic): find \"payload\": and then remainder
         * In production, use a real JSON parser. */
        char payload[MAX_LINE];
        payload[0] = '\0';
        const char *p = strstr(line, "\"payload\"");
        if (p) {
            const char *colon = strchr(p, ':');
            if (colon) {
                /* copy from colon+1 to end, but trim leading spaces */
                const char *q = colon + 1;
                while (*q == ' ' || *q == '\t') q++;
                strncpy(payload, q, sizeof(payload)-1);
                payload[sizeof(payload)-1] = '\0';
            }
        }

        if (strcmp(type, "health") == 0) {
            handle_health(id);
        }
        else if (strcmp(type, "exec") == 0) {
            /* Exec handled with timeout guard: spawn thread to avoid blocking main loop */
            /* For simplicity we'll handle synchronously here but plugin_timeout_seconds can be used
             * by host to kill long tasks. */
            handle_exec(id, payload);
        }
        else if (strcmp(type, "quit") == 0 || strcmp(type, "shutdown") == 0) {
            respond_ok(id, "{\"result\":\"shutting_down\"}");
            atomic_store(&running, false);
            break;
        }
        else {
            respond_error(id, 400, "unknown type");
        }
    }

    free(line);

    /* If loop ended due to EOF or running=false, shutdown gracefully */
    plugin_shutdown();

    /* Exit code: 0 for graceful, non-zero if external signal was reason for termination */
    return 0;
}
